proposals/simd/portable-simd.md

```diff
@@ -0,0 +1,1324 @@
+
+* `v32x4.store1(mem: Buffer, addr: ByteOffset, data: v128)`
+* `v32x4.store2(mem: Buffer, addr: ByteOffset, data: v128)`
+* `v32x4.store3(mem: Buffer, addr: ByteOffset, data: v128)`
+
+These functions store the first 1, 2, or 3 lanes to a buffer. They are only
+defined for the 4-lane interpretations.
+
+```python
+def partial_store(mem, addr, data, lanes):
+    for i in range(lanes):
+        mem.store(32, addr + i * 4, data[i])
+
+def v32x4.store1(mem, addr, data):
+    assert mem.in_range(addr, 4)
+    partial_store(mem, addr, data, 1)
+
+def v32x4.store2(mem, addr, data):
+    assert mem.in_range(addr, 8)
+    partial_store(mem, addr, data, 2)
+
+def v32x4.store3(mem, addr, data):
+    assert mem.in_range(addr, 12)
+    partial_store(mem, addr, data, 3)
+```
+
+## Floating-point sign bit operations
+
+These floating point operations are simple manipulations of the sign bit. No
+changes are made to the exponent or trailing significand bits, even for NaN
+inputs.
+
+### Negation
+* `f32x4.neg(a: v128) -> v128`
+* `f64x2.neg(a: v128) -> v128`
+
+Apply the IEEE `negate(x)` function to each lane. This simply inverts the sign
+bit, preserving all other bits.
+
+```python
+def S.neg(a):
+    return S.lanewise_unary(ieee.negate, a)
+```
+
+### Absolute value
+* `f32x4.abs(a: v128) -> v128`
+* `f64x2.abs(a: v128) -> v128`
+
+Apply the IEEE `abs(x)` function to each lane. This simply clears the sign bit,
+preserving all other bits.
+
+```python
+def S.abs(a):
+    return S.lanewise_unary(ieee.abs, a)
+```
+
+## Floating-point min and max
+
+These operations are not part of the IEEE 754-2008 standard. Notably, the
+`minNum` and `maxNum` operations defined here behave differently than the IEEE
+`minNum` and `maxNum` operations when one operand is a signaling NaN.
+
+The minimum and maximum value of +0 and -0 is computed as if -0 < +0.
+
+### NaN-propagating minimum
+* `f32x4.min(a: v128, b: v128) -> v128`
+* `f64x2.min(a: v128, b: v128) -> v128`
+
+Lane-wise minimum value, propagating NaNs:
+
+```python
+def S.min(a, b):
+    def min(x, y):
+        if isnan(x) or isnan(y):
+            return propagate_nan(x, y)
+        # Prefer -0 for min(-0, +0) and min(+0, -0).
+        if x == 0 and y == 0 and signbit(x) != signbit(y):
+            return -0.0
+        if x < y:
+            return x
+        else:
+            return y
+    return S.lanewise_binary(min, a, b)
+```
+
+### NaN-propagating maximum
+* `f32x4.max(a: v128, b: v128) -> v128`
+* `f64x2.max(a: v128, b: v128) -> v128`
+
+Lane-wise maximum value, propagating NaNs:
+
+```python
+def S.max(a, b):
+    def max(x, y):
+        if isnan(x) or isnan(y):
+            return propagate_nan(x, y)
+        # Prefer +0 for max(-0, +0) and max(+0, -0).
+        if x == 0 and y == 0 and signbit(x) != signbit(y):
+            return +0.0
+        if x > y:
+            return x
+        else:
+            return y
+    return S.lanewise_binary(max, a, b)
+```
+
+### NaN-suppressing minimum
```

Comment:
I agree, and you won't find these instruction in the table of proposed opcodes. I'll make a note that I omitted them.